draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,1630 @@
+=> [RubyToken::TkfLBRACK, PryStackExplorer, Psych::Visitors::DepthFirst]
+```
+
+So `initialize_matchable_scopes` will take all the modules it can find and try to create an expression out of it
+
+```Ruby
+  def initialize_matchable_scopes
+    scopes = ObjectSpace.each_object(Module).each_with_object([]) do |scope, aggregate|
+      expression = expression(scope) || next
+      # ... snip
+    end
+  end
+```
+
+With
+
+```Ruby
+  def expression(scope) # `scope` is one of the modules from ObjectSpace, e.g. Gem::Ext::BuildError, so a constant.
+    name = scope.name # redacted for simplicity.
+    # snip
+
+    config.expression_parser.try_parse(name)
+  end
+```
+
+`config.expression_parser` returns a `Mutant::Expression::Parser` by default. 
+
+`try_parse` will then return something like this
+
+```
+#<Mutant::Expression::Namespace::Exact scope_name="Gem::Ext::BuildError">
+```
+
+Ok, we went through a lot - let's look at the big picture again:
+
+```Ruby
+  def initialize_matchable_scopes
+    scopes = ObjectSpace.each_object(Module).each_with_object([]) do |scope, aggregate|
+      expression = expression(scope) || next
+      aggregate << Scope.new(scope, expression)
+    end
+
+    @matchable_scopes = scopes.sort_by { |scope| scope.expression.syntax }
+  end
+```
+
+* We're going through the full ObjectSpace
+* try to create an expression out of the given Module
+* if we can we store the module along side with it
+
+Now that we know how the initializer works we know half of the story of the call to 
+
+```Ruby
+Runner.call(
+  Env::Bootstrap.call(
+    call(arguments)
+  )
+).success?
+```
+
+But what happens when we call `Env::Bootstrap.call`? If you remember the `include` statement in the class declaration:
+
+```Ruby
+module Mutant
+  class Env
+    class Bootstrap
+      include Procto.call(:env)
+
+      def initialize(*)
+        # snip, we just discussed this
+      end
+    end
+  end
+end
+```
+
+you can see that we actually call `env` on our `Env::Bootstrap` object:
+
+```Ruby
+  def env
+    subjects = matched_subjects
+    Env.new(
+      actor_env:        Actor::Env.new(Thread),
+      config:           config,
+      integration:      @integration,
+      matchable_scopes: matchable_scopes,
+      mutations:        subjects.flat_map(&:mutations),
+      parser:           parser,
+      selector:         Selector::Expression.new(@integration),
+      subjects:         subjects
+    )
+  end
+```
+
+This now brings us to the topic of "subjects".
+
+### Subjects
+
+Let's go through the `env` method step by step:
+
+```Ruby
+  def env
+    subjects = matched_subjects
+    # snip
+  end
+```
+
+First we'll get all "matched subjects", which means all the subjects you want to mutation tests. I won't go into detail how the `matched_subjects` method works because that would probably deserve another blog post so let's just assume it returns your modules and classes you would like to mutation test. 
```

Comment:
> it returns your modules and classes you would like to mutation test

Not true, it returns the subjects, and each subject represents only a specific instance or singleton method that us later subject of mutation. Mutant does not yet mutate code at the class or module level.